File: app/src/main/python/ytmusic_recommender.py
"""
YTMusic Recommender - Uses ytmusicapi to get song-only recommendations
This module provides song search and recommendation functionality using YouTube Music API
"""

from ytmusicapi import YTMusic
import json
import logging

# Configure logging
logging.basicConfig(level=logging.INFO)
logger = logging.getLogger(__name__)


class YTMusicRecommender:
    """Handles YouTube Music API interactions for song recommendations"""
    
    def __init__(self):
        """Initialize YTMusic client"""
        try:
            self.yt = YTMusic()
            logger.info("YTMusic client initialized successfully")
        except Exception as e:
            logger.error(f"Failed to initialize YTMusic: {e}")
            self.yt = None
    
    def search_songs(self, query, limit=20):
        """
        Search for songs only (no videos, playlists, or albums)
        
        Args:
            query: Search query string
            limit: Maximum number of results (default 20)
            
        Returns:
            List of song dictionaries with keys: videoId, title, artist, album, duration, thumbnail
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return []
        
        try:
            results = self.yt.search(query, filter='songs', limit=limit)
            songs = []
            
            for item in results:
                song = {
                    'videoId': item.get('videoId', ''),
                    'title': item.get('title', 'Unknown Title'),
                    'artist': ', '.join([a['name'] for a in item.get('artists', [])]) or 'Unknown Artist',
                    'album': item.get('album', {}).get('name', 'Unknown Album') if item.get('album') else 'Unknown Album',
                    'duration': item.get('duration', '0:00'),
                    'thumbnail': item.get('thumbnails', [{}])[-1].get('url', '') if item.get('thumbnails') else '',
                    'videoType': item.get('videoType', 'MUSIC_VIDEO_TYPE_ATV')
                }
                
                # Only include audio tracks (songs), not music videos
                if song['videoType'] in ['MUSIC_VIDEO_TYPE_ATV', 'MUSIC_VIDEO_TYPE_OFFICIAL_SOURCE_MUSIC']:
                    songs.append(song)
            
            logger.info(f"Found {len(songs)} songs for query: {query}")
            return songs
            
        except Exception as e:
            logger.error(f"Search failed for query '{query}': {e}")
            return []
    
    def search_albums(self, query, limit=20):
        """
        Search for albums only
        
        Args:
            query: Search query string
            limit: Maximum number of results (default 20)
            
        Returns:
            List of album dictionaries with keys: browseId, title, artist, year, thumbnail
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return []
        
        try:
            results = self.yt.search(query, filter='albums', limit=limit)
            albums = []
            
            for item in results:
                album = {
                    'browseId': item.get('browseId', ''),
                    'title': item.get('title', 'Unknown Album'),
                    'artist': ', '.join([a['name'] for a in item.get('artists', [])]) or 'Unknown Artist',
                    'year': item.get('year', ''),
                    'thumbnail': item.get('thumbnails', [{}])[-1].get('url', '') if item.get('thumbnails') else '',
                    'type': item.get('resultType', 'album')
                }
                albums.append(album)
            
            logger.info(f"Found {len(albums)} albums for query: {query}")
            return albums
            
        except Exception as e:
            logger.error(f"Album search failed for query '{query}': {e}")
            return []
    
    def search_artists(self, query, limit=10):
        """
        Search for artists on YouTube Music
        
        Args:
            query: Search query string
            limit: Maximum number of results (default 10)
            
        Returns:
            List of artist dictionaries with browseId, name, thumbnail
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return []
        
        try:
            results = self.yt.search(query=query, filter='artists', limit=limit)
            
            if not results:
                logger.warning(f"No artists found for query: {query}")
                return []
            
            artists = []
            for item in results:
                artist = {
                    'browseId': item.get('browseId', ''),
                    'name': item.get('artist', 'Unknown Artist'),
                    'thumbnail': item.get('thumbnails', [{}])[-1].get('url', '') if item.get('thumbnails') else '',
                    'type': item.get('resultType', 'artist'),
                    'subscribers': item.get('subscribers', '')
                }
                artists.append(artist)
            
            logger.info(f"Found {len(artists)} artists for query: {query}")
            return artists
            
        except Exception as e:
            logger.error(f"Artist search failed for query '{query}': {e}")
            return []
    
    def get_album_details(self, browse_id):
        """
        Get album details including songs list
        
        Args:
            browse_id: Album browseId from search results
            
        Returns:
            Dictionary with album details and songs list
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return None
        
        try:
            album = self.yt.get_album(browseId=browse_id)
            
            if not album:
                logger.warning(f"No album details found for browseId: {browse_id}")
                return None
            
            songs = []
            for track in album.get('tracks', []):
                song = {
                    'videoId': track.get('videoId', ''),
                    'title': track.get('title', 'Unknown Title'),
                    'artist': ', '.join([a['name'] for a in track.get('artists', [])]) or 'Unknown Artist',
                    'album': album.get('title', 'Unknown Album'),
                    'duration': str(track.get('duration', '0:00')),
                    'thumbnail': track.get('thumbnails', [{}])[-1].get('url', '') if track.get('thumbnails') else ''
                }
                songs.append(song)
            
            result = {
                'browseId': browse_id,
                'title': album.get('title', 'Unknown Album'),
                'artist': ', '.join([a['name'] for a in album.get('artists', [])]) or 'Unknown Artist',
                'year': str(album.get('year', '')),
                'thumbnail': album.get('thumbnails', [{}])[-1].get('url', '') if album.get('thumbnails') else '',
                'trackCount': album.get('trackCount', len(songs)),
                'duration': album.get('duration', ''),
                'songs': songs
            }
            
            logger.info(f"Retrieved album details: {result['title']} with {len(songs)} songs")
            return result
            
        except Exception as e:
            logger.error(f"Failed to get album details for browseId '{browse_id}': {e}")
            return None
    
    def get_artist_details(self, browse_id):
        """
        Get artist details including top songs
        
        Args:
            browse_id: Artist browseId from search results
            
        Returns:
            Dictionary with artist details and songs list
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return None
        
        try:
            artist = self.yt.get_artist(channelId=browse_id)
            
            if not artist:
                logger.warning(f"No artist details found for browseId: {browse_id}")
                return None
            
            songs = []
            # Get songs from the artist (top tracks, singles, etc.)
            if 'songs' in artist and 'results' in artist['songs']:
                for track in artist['songs']['results'][:25]:  # Top 25 songs
                    song = {
                        'videoId': track.get('videoId', ''),
                        'title': track.get('title', 'Unknown Title'),
                        'artist': ', '.join([a['name'] for a in track.get('artists', [])]) or artist.get('name', 'Unknown Artist'),
                        'album': track.get('album', {}).get('name', '') if track.get('album') else '',
                        'duration': str(track.get('duration', '0:00')),
                        'thumbnail': track.get('thumbnails', [{}])[-1].get('url', '') if track.get('thumbnails') else ''
                    }
                    songs.append(song)
            
            result = {
                'browseId': browse_id,
                'name': artist.get('name', 'Unknown Artist'),
                'description': artist.get('description', ''),
                'thumbnail': artist.get('thumbnails', [{}])[-1].get('url', '') if artist.get('thumbnails') else '',
                'subscribers': artist.get('subscribers', ''),
                'songs': songs
            }
            
            logger.info(f"Retrieved artist details: {result['name']} with {len(songs)} songs")
            return result
            
        except Exception as e:
            logger.error(f"Failed to get artist details for browseId '{browse_id}': {e}")
            return None
    
    def get_song_recommendations(self, video_id, limit=25):
        """
        Get song recommendations based on a video ID
        Uses get_watch_playlist to get radio recommendations
        
        Args:
            video_id: YouTube video ID of a song
            limit: Maximum number of recommendations (default 25)
            
        Returns:
            List of recommended song dictionaries
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return []
        
        try:
            # Get watch playlist with radio parameter for recommendations
            watch_playlist = self.yt.get_watch_playlist(videoId=video_id, limit=limit, radio=True)
            
            if not watch_playlist or 'tracks' not in watch_playlist:
                logger.warning(f"No recommendations found for video_id: {video_id}")
                return []
            
            tracks = watch_playlist['tracks']
            recommendations = []
            
            for track in tracks:
                song = {
                    'videoId': track.get('videoId', ''),
                    'title': track.get('title', 'Unknown Title'),
                    'artist': ', '.join([a['name'] for a in track.get('artists', [])]) or 'Unknown Artist',
                    'album': track.get('album', {}).get('name', 'Unknown Album') if track.get('album') else 'Unknown Album',
                    'duration': str(track.get('length', '0:00')),
                    'thumbnail': track.get('thumbnail', [{}])[-1].get('url', '') if track.get('thumbnail') else '',
                    'videoType': track.get('videoType', 'MUSIC_VIDEO_TYPE_ATV')
                }
                
                # Filter: Only include audio tracks (songs), exclude music videos
                if song['videoType'] in ['MUSIC_VIDEO_TYPE_ATV', 'MUSIC_VIDEO_TYPE_OFFICIAL_SOURCE_MUSIC']:
                    recommendations.append(song)
            
            logger.info(f"Found {len(recommendations)}/{len(tracks)} song recommendations for video_id: {video_id}")
            return recommendations
            
        except Exception as e:
            logger.error(f"Failed to get recommendations for video_id '{video_id}': {e}")
            return []
    
    def get_recommendations_for_query(self, query, limit=25):
        """
        Get song recommendations based on a search query
        First searches for songs, then gets recommendations based on the first result
        
        Args:
            query: Search query string
            limit: Maximum number of recommendations (default 25)
            
        Returns:
            List of recommended song dictionaries
        """
        if not self.yt:
            logger.error("YTMusic client not initialized")
            return []
        
        try:
            # First, search for songs matching the query
            search_results = self.search_songs(query, limit=5)
            
            if not search_results:
                logger.warning(f"No songs found for query: {query}")
                return []
            
            # Use the first song to get recommendations
            first_song_id = search_results[0]['videoId']
            logger.info(f"Using song '{search_results[0]['title']}' (ID: {first_song_id}) for recommendations")
            
            recommendations = self.get_song_recommendations(first_song_id, limit)
            return recommendations
            
        except Exception as e:
            logger.error(f"Failed to get recommendations for query '{query}': {e}")
            return []


# Module-level functions for easy calling from Kotlin via Chaquopy

_recommender = None


def initialize():
    """Initialize the YTMusic recommender"""
    global _recommender
    try:
        _recommender = YTMusicRecommender()
        return "YTMusic initialized successfully"
    except Exception as e:
        logger.error(f"Initialization failed: {e}")
        return f"Initialization failed: {e}"


def search_songs(query, limit=20):
    """
    Search for songs
    Returns JSON string of song list
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        songs = _recommender.search_songs(query, limit)
        return json.dumps(songs)
    except Exception as e:
        logger.error(f"search_songs error: {e}")
        return json.dumps([])


def search_albums(query, limit=20):
    """
    Search for albums
    Returns JSON string of album list
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        albums = _recommender.search_albums(query, limit)
        return json.dumps(albums)
    except Exception as e:
        logger.error(f"search_albums error: {e}")
        return json.dumps([])


def search_artists(query, limit=10):
    """
    Search for artists
    Returns JSON string of artist list
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        artists = _recommender.search_artists(query, limit)
        return json.dumps(artists)
    except Exception as e:
        logger.error(f"search_artists error: {e}")
        return json.dumps([])


def get_song_recommendations(video_id, limit=25):
    """
    Get recommendations for a specific video ID
    Returns JSON string of recommendation list
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        recommendations = _recommender.get_song_recommendations(video_id, limit)
        return json.dumps(recommendations)
    except Exception as e:
        logger.error(f"get_song_recommendations error: {e}")
        return json.dumps([])


def get_recommendations_for_query(query, limit=25):
    """
    Get recommendations based on a search query
    Returns JSON string of recommendation list
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        recommendations = _recommender.get_recommendations_for_query(query, limit)
        return json.dumps(recommendations)
    except Exception as e:
        logger.error(f"get_recommendations_for_query error: {e}")
        return json.dumps([])


def get_album_details(browse_id):
    """
    Get album details including songs list
    Returns JSON string of album details with songs
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        album = _recommender.get_album_details(browse_id)
        return json.dumps(album) if album else json.dumps(None)
    except Exception as e:
        logger.error(f"get_album_details error: {e}")
        return json.dumps(None)


def get_artist_details(browse_id):
    """
    Get artist details including top songs
    Returns JSON string of artist details with songs
    """
    global _recommender
    if not _recommender:
        initialize()
    
    try:
        artist = _recommender.get_artist_details(browse_id)
        return json.dumps(artist) if artist else json.dumps(None)
    except Exception as e:
        logger.error(f"get_artist_details error: {e}")
        return json.dumps(None)

File: app/src/main/java/com/just_for_fun/synctax/util/YTMusicRecommender.kt
package com.just_for_fun.synctax.util

import android.util.Log
import com.chaquo.python.Python
import com.chaquo.python.PyException
import kotlinx.coroutines.*
import org.json.JSONArray
import org.json.JSONObject

/**
 * YouTube Music Recommender using ytmusicapi
 * Returns only songs (audio tracks), no music videos, playlists, or live streams
 */
object YTMusicRecommender {
    private const val TAG = "YTMusicRecommender"
    private val scope = CoroutineScope(Dispatchers.IO + SupervisorJob())
    
    @Volatile
    private var isInitialized = false
    
    /**
     * Initialize the Python ytmusicapi module
     * Call this once during app startup
     */
    fun initialize() {
        if (isInitialized) {
            Log.d(TAG, "Already initialized")
            return
        }
        
        scope.launch {
            try {
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val result = module.callAttr("initialize").toString()
                isInitialized = true
                Log.d(TAG, "Initialization result: $result")
            } catch (e: PyException) {
                Log.e(TAG, "Python initialization failed", e)
                isInitialized = false
            } catch (e: Exception) {
                Log.e(TAG, "Initialization failed", e)
                isInitialized = false
            }
        }
    }
    
    /**
     * Search for songs only (no videos, playlists, or albums)
     * @param query Search query
     * @param limit Maximum number of results (default 20)
     * @param onResult Callback with list of songs
     * @param onError Error callback
     */
    fun searchSongs(
        query: String,
        limit: Int = 20,
        onResult: (List<RecommendedSong>) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "searchSongs: query='$query', limit=$limit")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("search_songs", query, limit).toString()
                
                val songs = parseSongsFromJson(jsonResult)
                Log.d(TAG, "Found ${songs.size} songs for query: $query")
                
                withContext(Dispatchers.Main) {
                    onResult(songs)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python search_songs failed", e)
                withContext(Dispatchers.Main) {
                    onError("Search failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "searchSongs failed", e)
                withContext(Dispatchers.Main) {
                    onError("Search failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Search for albums
     * @param query Search query
     * @param limit Maximum number of results (default 20)
     * @param onResult Callback with list of albums
     * @param onError Error callback
     */
    fun searchAlbums(
        query: String,
        limit: Int = 20,
        onResult: (List<Album>) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "searchAlbums: query='$query', limit=$limit")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("search_albums", query, limit).toString()
                
                val albums = parseAlbumsFromJson(jsonResult)
                Log.d(TAG, "Found ${albums.size} albums for query: $query")
                
                withContext(Dispatchers.Main) {
                    onResult(albums)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python search_albums failed", e)
                withContext(Dispatchers.Main) {
                    onError("Album search failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "searchAlbums failed", e)
                withContext(Dispatchers.Main) {
                    onError("Album search failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Search for artists on YouTube Music
     * @param query Search query string
     * @param limit Maximum number of results (default 10)
     * @param onResult Callback with list of artists
     * @param onError Error callback
     */
    fun searchArtists(
        query: String,
        limit: Int = 10,
        onResult: (List<Artist>) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "searchArtists: query='$query', limit=$limit")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("search_artists", query, limit).toString()
                
                val artists = parseArtistsFromJson(jsonResult)
                Log.d(TAG, "Found ${artists.size} artists for query: $query")
                
                withContext(Dispatchers.Main) {
                    onResult(artists)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python search_artists failed", e)
                withContext(Dispatchers.Main) {
                    onError("Artist search failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "searchArtists failed", e)
                withContext(Dispatchers.Main) {
                    onError("Artist search failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Get song recommendations based on a YouTube video ID
     * Only returns audio tracks (songs), filters out music videos
     * @param videoId YouTube video ID of a song
     * @param limit Maximum number of recommendations (default 25)
     * @param onResult Callback with list of recommended songs
     * @param onError Error callback
     */
    fun getRecommendations(
        videoId: String,
        limit: Int = 25,
        onResult: (List<RecommendedSong>) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "getRecommendations: videoId='$videoId', limit=$limit")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("get_song_recommendations", videoId, limit).toString()
                
                val recommendations = parseSongsFromJson(jsonResult)
                Log.d(TAG, "Found ${recommendations.size} song recommendations for videoId: $videoId")
                
                recommendations.forEachIndexed { idx, song ->
                    Log.d(TAG, "Rec[$idx] id=${song.videoId} title='${song.title}' artist='${song.artist}'")
                }
                
                withContext(Dispatchers.Main) {
                    onResult(recommendations)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python get_song_recommendations failed", e)
                withContext(Dispatchers.Main) {
                    onError("Recommendations failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "getRecommendations failed", e)
                withContext(Dispatchers.Main) {
                    onError("Recommendations failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Get song recommendations based on a search query
     * First searches for songs, then gets recommendations based on the first result
     * @param query Search query
     * @param limit Maximum number of recommendations (default 25)
     * @param onResult Callback with list of recommended songs
     * @param onError Error callback
     */
    fun getRecommendationsForQuery(
        query: String,
        limit: Int = 25,
        onResult: (List<RecommendedSong>) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "getRecommendationsForQuery: query='$query', limit=$limit")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("get_recommendations_for_query", query, limit).toString()
                
                val recommendations = parseSongsFromJson(jsonResult)
                Log.d(TAG, "Found ${recommendations.size} recommendations for query: $query")
                
                withContext(Dispatchers.Main) {
                    onResult(recommendations)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python get_recommendations_for_query failed", e)
                withContext(Dispatchers.Main) {
                    onError("Query recommendations failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "getRecommendationsForQuery failed", e)
                withContext(Dispatchers.Main) {
                    onError("Query recommendations failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Get album details including songs list
     * @param browseId Album browseId from search results
     * @param onResult Callback with album details
     * @param onError Error callback
     */
    fun getAlbumDetails(
        browseId: String,
        onResult: (AlbumDetails?) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "getAlbumDetails: browseId='$browseId'")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("get_album_details", browseId).toString()
                
                val albumDetails = parseAlbumDetailsFromJson(jsonResult)
                Log.d(TAG, "Retrieved album: ${albumDetails?.title} with ${albumDetails?.songs?.size} songs")
                
                withContext(Dispatchers.Main) {
                    onResult(albumDetails)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python get_album_details failed", e)
                withContext(Dispatchers.Main) {
                    onError("Album details fetch failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "getAlbumDetails failed", e)
                withContext(Dispatchers.Main) {
                    onError("Album details fetch failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Get artist details including top songs
     * @param browseId Artist browseId from search results
     * @param onResult Callback with artist details
     * @param onError Error callback
     */
    fun getArtistDetails(
        browseId: String,
        onResult: (ArtistDetails?) -> Unit,
        onError: (String) -> Unit = { Log.e(TAG, it) }
    ) {
        scope.launch {
            try {
                Log.d(TAG, "getArtistDetails: browseId='$browseId'")
                
                val python = Python.getInstance()
                val module = python.getModule("ytmusic_recommender")
                val jsonResult = module.callAttr("get_artist_details", browseId).toString()
                
                val artistDetails = parseArtistDetailsFromJson(jsonResult)
                Log.d(TAG, "Retrieved artist: ${artistDetails?.name} with ${artistDetails?.songs?.size} songs")
                
                withContext(Dispatchers.Main) {
                    onResult(artistDetails)
                }
            } catch (e: PyException) {
                Log.e(TAG, "Python get_artist_details failed", e)
                withContext(Dispatchers.Main) {
                    onError("Artist details fetch failed: ${e.message}")
                }
            } catch (e: Exception) {
                Log.e(TAG, "getArtistDetails failed", e)
                withContext(Dispatchers.Main) {
                    onError("Artist details fetch failed: ${e.message}")
                }
            }
        }
    }
    
    /**
     * Parse songs from JSON string returned by Python
     */
    private fun parseSongsFromJson(jsonString: String): List<RecommendedSong> {
        val songs = mutableListOf<RecommendedSong>()
        
        try {
            val jsonArray = JSONArray(jsonString)
            
            for (i in 0 until jsonArray.length()) {
                val jsonObject = jsonArray.getJSONObject(i)
                
                val videoId = jsonObject.optString("videoId", "")
                if (videoId.isEmpty()) continue
                
                val song = RecommendedSong(
                    title = jsonObject.optString("title", "Unknown Title"),
                    artist = jsonObject.optString("artist", "Unknown Artist"),
                    videoId = videoId,
                    watchUrl = "https://www.youtube.com/watch?v=$videoId",
                    thumbnail = jsonObject.optString("thumbnail", "")
                )
                
                songs.add(song)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse songs JSON", e)
        }
        
        return songs
    }
    
    /**
     * Parse albums from JSON string returned by Python
     */
    private fun parseAlbumsFromJson(jsonString: String): List<Album> {
        val albums = mutableListOf<Album>()
        
        try {
            val jsonArray = JSONArray(jsonString)
            
            for (i in 0 until jsonArray.length()) {
                val jsonObject = jsonArray.getJSONObject(i)
                
                val browseId = jsonObject.optString("browseId", "")
                if (browseId.isEmpty()) continue
                
                val album = Album(
                    browseId = browseId,
                    title = jsonObject.optString("title", "Unknown Album"),
                    artist = jsonObject.optString("artist", "Unknown Artist"),
                    year = jsonObject.optString("year", ""),
                    thumbnail = jsonObject.optString("thumbnail", "")
                )
                
                albums.add(album)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse albums JSON", e)
        }
        
        return albums
    }
    
    private fun parseArtistsFromJson(jsonString: String): List<Artist> {
        val artists = mutableListOf<Artist>()
        
        try {
            val jsonArray = JSONArray(jsonString)
            
            for (i in 0 until jsonArray.length()) {
                val jsonObject = jsonArray.getJSONObject(i)
                
                val browseId = jsonObject.optString("browseId", "")
                if (browseId.isEmpty()) continue
                
                val artist = Artist(
                    browseId = browseId,
                    name = jsonObject.optString("name", "Unknown Artist"),
                    thumbnail = jsonObject.optString("thumbnail", ""),
                    subscribers = jsonObject.optString("subscribers", "")
                )
                
                artists.add(artist)
            }
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse artists JSON", e)
        }
        
        return artists
    }
    
    private fun parseAlbumDetailsFromJson(jsonString: String): AlbumDetails? {
        try {
            if (jsonString == "null" || jsonString.isEmpty()) return null
            
            val jsonObject = JSONObject(jsonString)
            val songsArray = jsonObject.optJSONArray("songs") ?: JSONArray()
            
            val songs = mutableListOf<RecommendedSong>()
            for (i in 0 until songsArray.length()) {
                val songObj = songsArray.getJSONObject(i)
                val videoId = songObj.optString("videoId", "")
                val song = RecommendedSong(
                    videoId = videoId,
                    title = songObj.optString("title", "Unknown Title"),
                    artist = songObj.optString("artist", "Unknown Artist"),
                    thumbnail = songObj.optString("thumbnail", ""),
                    watchUrl = "https://www.youtube.com/watch?v=$videoId"
                )
                songs.add(song)
            }
            
            return AlbumDetails(
                browseId = jsonObject.optString("browseId", ""),
                title = jsonObject.optString("title", "Unknown Album"),
                artist = jsonObject.optString("artist", "Unknown Artist"),
                year = jsonObject.optString("year", ""),
                thumbnail = jsonObject.optString("thumbnail", ""),
                trackCount = jsonObject.optInt("trackCount", songs.size),
                duration = jsonObject.optString("duration", ""),
                songs = songs
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse album details JSON", e)
            return null
        }
    }
    
    private fun parseArtistDetailsFromJson(jsonString: String): ArtistDetails? {
        try {
            if (jsonString == "null" || jsonString.isEmpty()) return null
            
            val jsonObject = JSONObject(jsonString)
            val songsArray = jsonObject.optJSONArray("songs") ?: JSONArray()
            
            val songs = mutableListOf<RecommendedSong>()
            for (i in 0 until songsArray.length()) {
                val songObj = songsArray.getJSONObject(i)
                val videoId = songObj.optString("videoId", "")
                val song = RecommendedSong(
                    videoId = videoId,
                    title = songObj.optString("title", "Unknown Title"),
                    artist = songObj.optString("artist", "Unknown Artist"),
                    thumbnail = songObj.optString("thumbnail", ""),
                    watchUrl = "https://www.youtube.com/watch?v=$videoId"
                )
                songs.add(song)
            }
            
            return ArtistDetails(
                browseId = jsonObject.optString("browseId", ""),
                name = jsonObject.optString("name", "Unknown Artist"),
                description = jsonObject.optString("description", ""),
                thumbnail = jsonObject.optString("thumbnail", ""),
                subscribers = jsonObject.optString("subscribers", ""),
                songs = songs
            )
        } catch (e: Exception) {
            Log.e(TAG, "Failed to parse artist details JSON", e)
            return null
        }
    }
}

/**
 * Data class for album information
 */
data class Album(
    val browseId: String,
    val title: String,
    val artist: String,
    val year: String,
    val thumbnail: String
)

/**
 * Data class for artist information
 */
data class Artist(
    val browseId: String,
    val name: String,
    val thumbnail: String,
    val subscribers: String
)

/**
 * Data class for album details with songs
 */
data class AlbumDetails(
    val browseId: String,
    val title: String,
    val artist: String,
    val year: String,
    val thumbnail: String,
    val trackCount: Int,
    val duration: String,
    val songs: List<RecommendedSong>
)

/**
 * Data class for artist details with songs
 */
data class ArtistDetails(
    val browseId: String,
    val name: String,
    val description: String,
    val thumbnail: String,
    val subscribers: String,
    val songs: List<RecommendedSong>
)
